tempesta_fw/cache.c

```diff
@@ -1847,54 +1886,138 @@ tfw_cache_purge_method(TfwHttpReq *req)
 		tfw_http_send_resp(req, 200, "purge: success");
 }
 
+/**
+ * Add page from cache into response. Different strategies are used to
+ * avoid extra data copying depending on client connection type:
+ * - for http connections - pages are reused in skbs and SKBTX_SHARED_FRAG is
+ * set to avoid any data copies.
+ * - for https connections - pages are reused in skbs and SKBTX_SHARED_FRAG is
+ * set, but in-place crypto operations are not allowed, so data copy happens
+ * right before data is pushed into network.
+ * - for h2 connections - copying happens on constructing responses from
+ * cache, and SKBTX_SHARED_FRAG is left unset allowing in-place crypto
+ * operations.
+ *
+ * Since we can't encrypt shared data in-place we always copy it, so we need
+ * reserve some space in cached pages to avoid extra skb fragmentation. To do
+ * so, we reserve space for h2 frame header only before first body fragment,
+ * since it shares page with cached headers. The other body chunks begin at
+ * beginning of the cache record page right after cache record header. Since
+ * cache record header is always bigger than h2 frame header there is no need
+ * to reserve extra space for h2 frame header.
+ */
+static int
+tfw_cache_add_body_page(TfwMsgIter *it, char *p, int sz, TfwFrameHdr *frame_hdr,
+			bool h2, bool *first_frag, bool last_frag)
+{
+	int off;
+	struct page *page;
+
+	/*
+	 * @sz is guarantied to be bigger than FRAME_HEADER_SIZE if frame header
+	 * is stored in cache before data. True for first fragment.
+	 */
+	if (h2) {
+		char *new_p;
+		if (!(page = alloc_page(GFP_ATOMIC))) {
+			return -ENOMEM;
+		}
+		new_p = page_address(page);
+		off = 0;
+		frame_hdr->flags = last_frag ? HTTP2_F_END_STREAM : 0;
+		if (*first_frag) {
+			frame_hdr->length = sz - FRAME_HEADER_SIZE;
+			memcpy_fast(new_p, p, sz);
```

Comment:
Why should we reserve the space for FRAME_HEADER_SIZE in the cache page for first H2 frame - in the `tfw_cache_h2_copy_body()`?
In both cases (first frame and subsequent frames) we create new page and copy into it the next body part from the cache page with the frame header insertion. So, the difference between these two cases is not clear for me.

Reply:
Reviewed that and now I totally agree here.  now the same approach is used for both cases.